### ISSUE TITLE
CMake-based build fails in git worktrees

### ISSUE THREAD

**Issue by nbraud** [Z3Prover/z3#1227]
The CMake-based build system assumes that if `.git` exists, it is a directory, which isn't always the case:

```
 ~/everest/z3-injectivity/build $ cmake ..        
-- The CXX compiler identification is GNU 6.3.0                                            
-- Check for working CXX compiler: /usr/bin/c++                                            
-- Check for working CXX compiler: /usr/bin/c++ -- works                                   
-- Detecting CXX compiler ABI info                                                         
-- Detecting CXX compiler ABI info - done                                                  
-- Detecting CXX compile features                                                          
-- Detecting CXX compile features - done                                                   
-- Z3 version 4.5.1.0                                                                      
CMake Error at cmake/git_utils.cmake:20 (message):                                         
  GIT_DIR ("/home/t-nibrau/everest/z3-injectivity/.git") is not a directory                
Call Stack (most recent call first):                                                       
  CMakeLists.txt:102 (add_git_dir_dependency)                                              
                                                                                           
                                                                                           
-- Configuring incomplete, errors occurred!                                                
See also "/home/t-nibrau/everest/z3-injectivity/build/CMakeFiles/CMakeOutput.log".         

 ~/everest/z3-injectivity/build $ ll ../.git      
-rw-r--r-- 1 t-nibrau t-nibrau 64 Aug 16 17:21 ../.git

 ~/everest/z3-injectivity/build $ cat ../.git     
gitdir: /home/t-nibrau/everest/z3/.git/worktrees/z3-injectivity
```

**Comment by delcypher** [comment]
You can probably workaround this by passing `-DINCLUDE_GIT_DESCRIBE=OFF` and `-DINCLUDE_GIT_HASH=OFF` to CMake.

Your shell output is kind of hard to read but it looks like `.git` in your case in a plain file that contains text that points to a different tree. How did you create your git clone? If I can reproduce it I can probably fix it.

**Comment by nbraud** [comment]
Sorry, I didn't realise the prompt was somewhat in the way; I just edited it out.
Yes, `.git` is indeed a plain file that only refers to another `.git` tree.

Here is how to reproduce it:
```
/tmp/z3-cmake $ git clone https://github.com/Z3Prover/z3    
Cloning into 'z3'...                                                           
remote: Counting objects: 58704, done.                                         
remote: Compressing objects: 100% (39/39), done.                               
remote: Total 58704 (delta 37), reused 44 (delta 25), pack-reused 58640        
Receiving objects: 100% (58704/58704), 23.09 MiB | 17.37 MiB/s, done.          
Resolving deltas: 100% (46932/46932), done.                                    

/tmp/z3-cmake $ git -C z3/ worktree add ../z3-worktree      
Preparing ../z3-worktree (identifier z3-worktree)                              
HEAD is now at dca30ab2 Merge pull request #1225 from nbraud/nbraud/injectivity

/tmp/z3-cmake $ cat z3-worktree/.git                        
gitdir: /tmp/z3-cmake/z3/.git/worktrees/z3-worktree                            
```

**Comment by delcypher** [comment]
This should be fixed by #1228 .

**Comment by delcypher** [comment]
@nbraud Thanks for introducing me to git worktrees too. I didn't know about this feature and it's pretty neat :)

**Comment by nbraud** [comment]
I can confirm it's fixed in #1228.

@delcypher You are welcome, and thanks for solving the issue.  :)
I find it very convenient too, since I usually have a bunch of different branches and builds of Z3 around; the only pain point is that any given branch can only be checked out in one worktree at a time (for obvious reasons: git would need to apply changes commited in one checkout to the other, which would be confusing)
